File: examples/harness-typescript/readThroughCache.ts
// Minimal dnr read-through cache adapter for TypeScript/Node agent harnesses.
// dnr misses are soft, normal parsing still happens, and cache writes never
// fail the user task.

import { spawn } from "node:child_process";
import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";

type Reader = (file: string) => Promise<string>;

type RunResult = {
  code: number | null;
  stdout: string;
  stderr: string;
};

function dnrBin(): string {
  return process.env.DNR_BIN || "dnr";
}

function runDnr(args: string[]): Promise<RunResult | null> {
  return new Promise((resolve) => {
    const child = spawn(dnrBin(), args, { stdio: ["ignore", "pipe", "pipe"] });
    let stdout = "";
    let stderr = "";

    child.stdout.setEncoding("utf8");
    child.stderr.setEncoding("utf8");
    child.stdout.on("data", (chunk) => {
      stdout += chunk;
    });
    child.stderr.on("data", (chunk) => {
      stderr += chunk;
    });
    child.on("error", () => resolve(null));
    child.on("close", (code) => resolve({ code, stdout, stderr }));
  });
}

export async function readCachedTranscript(file: string): Promise<string | null> {
  const result = await runDnr(["read", file]);
  if (!result || result.code !== 0) return null;
  return result.stdout.trim().length > 0 ? result.stdout : null;
}

export async function cacheTranscript(
  file: string,
  transcript: string,
  options: { method: string; transcriber: string },
): Promise<void> {
  if (transcript.trim().length === 0) return;

  let dir: string | null = null;
  try {
    dir = await mkdtemp(join(tmpdir(), "dnr-"));
    const transcriptFile = join(dir, "transcript.md");
    await writeFile(transcriptFile, transcript, "utf8");
    await runDnr([
      "record",
      file,
      "--transcript-file",
      transcriptFile,
      "--method",
      options.method,
      "--transcriber",
      options.transcriber,
    ]);
  } catch {
    // Cache writes are best-effort. Never fail the user's task.
  } finally {
    if (dir) await rm(dir, { recursive: true, force: true });
  }
}

export async function readWithDnrCache(
  file: string,
  normalReader: Reader,
  options: { method: string; transcriber: string },
): Promise<string> {
  const cached = await readCachedTranscript(file);
  if (cached !== null) return cached;

  const transcript = await normalReader(file);
  await cacheTranscript(file, transcript, options);
  return transcript;
}

export async function queryFolder(folder: string, queryArgs: string[]): Promise<string | null> {
  await runDnr(["index", folder]);
  const result = await runDnr(["query", folder, ...queryArgs]);
  if (!result || result.code !== 0) return null;
  return result.stdout;
}

File: examples/harness-python/read_through_cache.py
"""Minimal dnr read-through cache adapter for Python agent harnesses.

Copy the shape, not necessarily this exact file. The important behavior:
dnr misses are soft, normal parsing still happens, and cache writes never fail
the user task.
"""
from __future__ import annotations

from collections.abc import Callable
from pathlib import Path
import shutil
import subprocess
import tempfile


Parser = Callable[[Path], str]


def _dnr_bin() -> str | None:
    return shutil.which("dnr")


def _run(args: list[str]) -> subprocess.CompletedProcess[str] | None:
    dnr = _dnr_bin()
    if not dnr:
        return None
    try:
        return subprocess.run(
            [dnr, *args],
            check=False,
            text=True,
            encoding="utf-8",
            errors="replace",
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
    except OSError:
        return None


def read_cached_transcript(path: str | Path) -> str | None:
    """Return a verified cached transcript, or None on any miss."""
    result = _run(["read", str(Path(path))])
    if result is None or result.returncode != 0:
        return None
    transcript = result.stdout
    return transcript if transcript.strip() else None


def cache_transcript(
    path: str | Path,
    transcript: str,
    *,
    method: str,
    transcriber: str,
) -> None:
    """Best-effort cache write. Never raise into the user task."""
    if not transcript.strip():
        return
    transcript_file: Path | None = None
    try:
        with tempfile.NamedTemporaryFile("w", encoding="utf-8", suffix=".md", delete=False) as tmp:
            tmp.write(transcript)
            transcript_file = Path(tmp.name)
        _run(
            [
                "record",
                str(Path(path)),
                "--transcript-file",
                str(transcript_file),
                "--method",
                method,
                "--transcriber",
                transcriber,
            ]
        )
    except OSError:
        return
    finally:
        if transcript_file is not None:
            try:
                transcript_file.unlink(missing_ok=True)
            except OSError:
                pass


def read_with_dnr_cache(
    path: str | Path,
    normal_reader: Parser,
    *,
    method: str,
    transcriber: str,
) -> str:
    """Use dnr if present; otherwise fall back to the harness's normal reader."""
    p = Path(path)
    cached = read_cached_transcript(p)
    if cached is not None:
        return cached

    transcript = normal_reader(p)
    cache_transcript(p, transcript, method=method, transcriber=transcriber)
    return transcript


def query_folder(folder: str | Path, *query_args: str) -> str | None:
    """Index, then query cached folder transcripts. Returns stdout or None on soft miss."""
    _run(["index", str(Path(folder))])
    result = _run(["query", str(Path(folder)), *query_args])
    if result is None or result.returncode != 0:
        return None
    return result.stdout
